tasks.c

```diff
@@ -3734,6 +3734,21 @@ void vTaskStartScheduler( void )
      * from getting optimized out as it is no longer used by the kernel. */
     ( void ) uxTopUsedPriority;
 
+    /* clean up so vTaskStartScheduler can be called again */
```

Comment:
We have a series of tests that I'd like to run from a clean system state each time, using catch2 or some other test framework. This consists of starting the system up, using an outside thread to interact with the system, and then shutting it down. With the present POSIX port not shutting down cleanly or restarting cleanly we aren't able to perform these tests in a clean environment, within the same process.

Reply:
@chinglee-iot we've considered those couple of cases but they seemed more complex. Additionally, because FreeRTOS is leaving so many resources behind it makes using valgrind more difficult. For our single scheduler internal use case we had a dozen suppressions for valgrind so we could tell it to ignore the various memory leaks. We also had to force shutdown with a call to exit() due to the shutdown hang issues.

Imo  you'd send a strong message if FreeRTOS could clean up after itself, even if only in the POSIX port case, in terms of helping users be able to run tools like valgrind and not having to force exit() simulated apps.

Given that it isn't a typical use case it was always my intention to support it without significant, or any if possible, impact on the kernel proper.

Aren't we close at this point to doing so? I'm not asking for the kernel devs to do the work to add support for it, I'm happy to continue with this PR until its ready for inclusion.

Reply:
@chinglee-iot if you'd like me to update the PR with any changes, for example if you'd like me to update the PR to use the internal list (to avoid needing trace enabled) just let me know and I'll rebase the trace commit to use that approach instead. I'd like to do what I can to keep this PR moving as I think it would help others when using the posix port.

Reply:
@chinglee-iot ok, let me split the PR into two parts and then apply the list changes to the cleanup side if it helps to keep things moving.